### ISSUE TITLE
Problem with yml configuration file parsing map

### ISSUE THREAD

**Issue by yygnay** [spring-projects/spring-boot#13404]
Spring boot version： 2.0.2 (1.5.13 is normal)

Problem Description：

If the map key is with “/”, it cannot be parsed normally. Others are normal.

Example：

```
application.yml 

shiro:
  testMap:
    /test1: test1
    /test2: test2
```

Configuration class 

```java
@Data
@ConfigurationProperties(prefix="shiro")
public class ShiroProperties {
	private Map<String,String> testMap;
}
```

Test class

```java
@Transactional
@RunWith(SpringRunner.class)
@SpringBootTest(classes=DataManage.class)
public class ShiroTest extends BaseTransactionalTest{

	@Autowired
	private ShiroProperties shiroConfig;
	
	@Test
	public void read() {
		System.out.println(shiroConfig.toString());
	}
	
}
```

console info：

```
ShiroProperties(testMap={test1=test1, test2=test2})
```

**Comment by philwebb** [comment]
I'm wonder if there's any way we can make this more user friendly. I've flagged the issue for discussion at our next call.

**Comment by yygnay** [comment]
Thank you! According to your suggestion it can be used normally, I hope there are relevant examples in the official document...

**Comment by mbhave** [comment]
We might be able to get the original key in the `EntryBinder`.

**Comment by edysli** [comment]
This also tripped me up while migrating to 2.0 because I have URLs as keys, something like:

    myapp
      serviceUrls:
        'https://example.org/test': 'https://test.example.org/Endpoint'

and it used to work fine with 1.5. After finding this issue, I changed this map to:

    '[https://example.org/test]': 'https://test.example.org/Endpoint'

and now it parses correctly. Please document this explicitly in the migration guide or reference doc.

**Comment by snicoll** [comment]
@edysli there is already an issue for that #13506

**Comment by edysli** [comment]
@snicoll Yeah I saw it, I just didn't know which one was best to host my "me too" example so I picked the lowest id... :wink: I can move my comment over to #13506, if that's less messy.

**Comment by Chatom** [comment]
Hi, it indeed works well with the use of brackets, but it is quite unnatural to do so.
I'd understand the use of quotes (even if they're not necessary in yaml syntax) but brackets seems a bit too much.
Another problem is that the mapping to the cleaned string without any warning can easily lead to errors if the config is not checked properly...

**Comment by wilkinsona** [comment]
@Chatom We're aware that there's some room for improvement here. That's why this issue remains open.

**Comment by maraswrona** [comment]
Any idea when this could be improved? Is there some better way around this (other than using brackets) ?

**Comment by wilkinsona** [comment]
@maraswrona The issue is in the 2.x milestone which means it'll be considered for inclusion in 2.3.

**Comment by kdowbecki** [comment]
I'd really prefer the binder to fail fast with an exception when it discovers uncapped property names. Currently 2.2.1 silently modifies the property name by removing this characters which is a nightmare to debug.

**Comment by pablocavalieri** [comment]
> The new binder is much stricter about property names which means you need to surround them in square brackets. Try the following:
> 
> ```yaml
> shiro:
>   testMap:
>     "[/test1]": test1
>     "[/test2]": test2
> ```

Is there a way to scape the square brackets? I need my Map key to be `some value[value between square brackets]`

**Comment by philwebb** [comment]
@pablocavalieri I believe you can do this:

```yaml
testMap:
  "[some value[value between square brackets]]": "test"
```

**Comment by minhduyphamdev** [comment]
2024 any new update, I meet an issue when use underscore, the spring auto escape it and map same value with the key with no underscore.

**Comment by philwebb** [comment]
@minhduyphamdev There's no update as yet. Please subscribe to the issue if you want to keep informed.
